modules/images/image-loading-optimization/detect.js

```diff
@@ -0,0 +1,279 @@
+		}
+	} );
+
+	/** @type {IntersectionObserverEntry[]} */
+	const elementIntersections = [];
+
+	/** @type {?IntersectionObserver} */
+	let intersectionObserver;
+
+	function disconnectIntersectionObserver() {
+		if ( intersectionObserver instanceof IntersectionObserver ) {
+			intersectionObserver.disconnect();
+			win.removeEventListener( 'scroll', disconnectIntersectionObserver ); // Clean up, even though this is registered with once:true.
+		}
+	}
+
+	// Wait for the intersection observer to report back on the initially-visible elements.
+	// Note that the first callback will include _all_ observed entries per <https://github.com/w3c/IntersectionObserver/issues/476>.
+	if ( breadcrumbedElementsMap.size > 0 ) {
+		await new Promise( ( resolve ) => {
+			intersectionObserver = new IntersectionObserver(
+				( entries ) => {
+					for ( const entry of entries ) {
+						if ( entry.isIntersecting ) {
+							elementIntersections.push( entry );
+						}
+					}
+					resolve();
+				},
+				{
+					root: null, // To watch for intersection relative to the device's viewport.
+					threshold: 0.0, // As soon as even one pixel is visible.
+				}
+			);
+
+			for ( const element of breadcrumbedElementsMap.keys() ) {
+				if ( ! adminBar || ! adminBar.contains( element ) ) {
+					intersectionObserver.observe( element );
+				}
+			}
+		} );
+
+		// Stop observing as soon as the page scrolls since we only want initial-viewport elements.
+		win.addEventListener( 'scroll', disconnectIntersectionObserver, {
+			once: true,
+			passive: true,
+		} );
+	}
+
+	// TODO: Use a local copy of web-vitals.
+	const { onLCP } = await import(
+		// eslint-disable-next-line import/no-unresolved
+		'https://unpkg.com/web-vitals@3/dist/web-vitals.js?module'
+	);
+
+	/** @type {LCPMetric[]} */
+	const lcpMetricCandidates = [];
+
+	// Obtain at least one LCP candidate. More may be reported before the page finishes loading.
+	await new Promise( ( resolve ) => {
+		onLCP(
+			( metric ) => {
+				lcpMetricCandidates.push( metric );
+				resolve();
+			},
+			{
+				// This avoids needing to click to finalize LCP candidate. While this is helpful for testing, it also
+				// ensures that we always get an LCP candidate reported. Otherwise, the callback may never fire if the
+				// user never does a click or keydown, per <https://github.com/GoogleChrome/web-vitals/blob/07f6f96/src/onLCP.ts#L99-L107>.
+				reportAllChanges: true,
+			}
+		);
+	} );
+
+	// Wait until the resources on the page have fully loaded.
+	await new Promise( ( resolve ) => {
+		if ( doc.readyState === 'complete' ) {
+			resolve();
+		} else {
+			win.addEventListener( 'load', resolve, { once: true } );
+		}
+	} );
+
+	// Stop observing.
+	disconnectIntersectionObserver();
+	if ( isDebug ) {
+		log( 'Detection is stopping.' );
+	}
+
+	/** @type {PageMetrics} */
+	const pageMetrics = {
+		viewport: {
+			width: win.innerWidth,
+			height: win.innerHeight,
+		},
+		elements: [],
+	};
```

Comment:
Curious how you're thinking of using various viewport dimensions in the data when applying optimizations?

Reply:
If it turns out that that there are different LCP images for different viewports, then `fetchpriority` would not be added to the `img` elements themselves. Instead the `viewport` widths captured here would then be used to construct the `media` attribute of `preload` links so that only the relevant image URL is prioritized for the given viewport. In this way, the same markup can be served to both desktop and mobile, while both still get optimized.